File: sdk/elixir/lib/rocksky/library.ex
defmodule Rocksky.Library do
  @moduledoc """
  Ergonomic wrappers for the authenticated `app.rocksky.library.*` API — the
  Subsonic/navidrome-compatible surface over a user's uploaded music.

  Every function requires a non-empty `token` (Bearer). Optional parameters go
  in an `opts` map with camelCase string keys; `base` overrides the AppView
  URL. Returns `{:ok, value}` | `{:error, message}`.
  """

  @doc "app.rocksky.library.ping — requires auth."
  def ping(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.ping", %{})

  @doc "app.rocksky.library.getLicense — requires auth."
  def get_license(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getLicense", %{})

  @doc "app.rocksky.library.getMusicFolders — requires auth."
  def get_music_folders(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getMusicFolders", %{})

  @doc "app.rocksky.library.getScanStatus — requires auth."
  def get_scan_status(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getScanStatus", %{})

  @doc "app.rocksky.library.startScan — requires auth."
  def start_scan(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.startScan", %{})

  @doc "app.rocksky.library.getUser — requires auth."
  def get_user(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getUser", %{})

  @doc "app.rocksky.library.getArtists — requires auth."
  def get_artists(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getArtists", %{})

  @doc "app.rocksky.library.getIndexes — requires auth."
  def get_indexes(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getIndexes", %{})

  @doc "app.rocksky.library.getArtist — requires auth."
  def get_artist(token, id, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getArtist", %{"id" => id})

  @doc "app.rocksky.library.getArtistInfo — requires auth."
  def get_artist_info(token, id, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getArtistInfo", %{"id" => id})

  @doc "app.rocksky.library.getAlbum — requires auth."
  def get_album(token, id, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getAlbum", %{"id" => id})

  @doc "app.rocksky.library.getAlbumList — requires auth."
  def get_album_list(token, type, opts \\ %{}, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getAlbumList", Map.merge(%{"type" => type}, opts))

  @doc "app.rocksky.library.getAlbumInfo — requires auth."
  def get_album_info(token, id, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getAlbumInfo", %{"id" => id})

  @doc "app.rocksky.library.getSong — requires auth."
  def get_song(token, id, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getSong", %{"id" => id})

  @doc "app.rocksky.library.getRandomSongs — requires auth."
  def get_random_songs(token, opts \\ %{}, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getRandomSongs", opts)

  @doc "app.rocksky.library.getSongsByGenre — requires auth."
  def get_songs_by_genre(token, genre, opts \\ %{}, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getSongsByGenre", Map.merge(%{"genre" => genre}, opts))

  @doc "app.rocksky.library.getSimilarSongs — requires auth."
  def get_similar_songs(token, id, opts \\ %{}, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getSimilarSongs", Map.merge(%{"id" => id}, opts))

  @doc "app.rocksky.library.getTopSongs — requires auth."
  def get_top_songs(token, artist, opts \\ %{}, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getTopSongs", Map.merge(%{"artist" => artist}, opts))

  @doc "app.rocksky.library.getLyrics — requires auth."
  def get_lyrics(token, opts \\ %{}, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getLyrics", opts)

  @doc "app.rocksky.library.getMusicDirectory — requires auth."
  def get_music_directory(token, id, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getMusicDirectory", %{"id" => id})

  @doc "app.rocksky.library.getGenres — requires auth."
  def get_genres(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getGenres", %{})

  @doc "app.rocksky.library.search — requires auth."
  def search(token, query, opts \\ %{}, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.search", Map.merge(%{"query" => query}, opts))

  @doc "app.rocksky.library.getStarred — requires auth."
  def get_starred(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getStarred", %{})

  @doc "app.rocksky.library.star — requires auth."
  def star(token, id, opts \\ %{}, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.star", Map.merge(%{"id" => id}, opts))

  @doc "app.rocksky.library.unstar — requires auth."
  def unstar(token, id, opts \\ %{}, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.unstar", Map.merge(%{"id" => id}, opts))

  @doc "app.rocksky.library.getPlaylists — requires auth."
  def get_playlists(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getPlaylists", %{})

  @doc "app.rocksky.library.getPlaylist — requires auth."
  def get_playlist(token, id, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getPlaylist", %{"id" => id})

  @doc "app.rocksky.library.createPlaylist — requires auth."
  def create_playlist(token, name, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.createPlaylist", %{"name" => name})

  @doc "app.rocksky.library.updatePlaylist — requires auth."
  def update_playlist(token, playlist_id, opts \\ %{}, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.updatePlaylist", Map.merge(%{"playlistId" => playlist_id}, opts))

  @doc "app.rocksky.library.deletePlaylist — requires auth."
  def delete_playlist(token, id, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.deletePlaylist", %{"id" => id})

  @doc "app.rocksky.library.deleteSong — requires auth."
  def delete_song(token, id, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.deleteSong", %{"id" => id})

  @doc "app.rocksky.library.deleteAlbum — requires auth."
  def delete_album(token, id, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.deleteAlbum", %{"id" => id})

  @doc "app.rocksky.library.scrobble — requires auth."
  def scrobble(token, id, opts \\ %{}, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.scrobble", Map.merge(%{"id" => id}, opts))

  @doc "app.rocksky.library.updateNowPlaying — requires auth."
  def update_now_playing(token, id, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.updateNowPlaying", %{"id" => id})

  @doc "app.rocksky.library.getNowPlaying — requires auth."
  def get_now_playing(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getNowPlaying", %{})

  @doc "app.rocksky.library.getPlayQueue — requires auth."
  def get_play_queue(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getPlayQueue", %{})

  @doc "app.rocksky.library.savePlayQueue — requires auth."
  def save_play_queue(token, opts \\ %{}, base \\ ""),
    do: :rocksky.library_post(base, token, "app.rocksky.library.savePlayQueue", opts)

  @doc "app.rocksky.library.getStreamUrl — requires auth."
  def get_stream_url(token, id, opts \\ %{}, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getStreamUrl", Map.merge(%{"id" => id}, opts))

  @doc "app.rocksky.library.getDownloadUrl — requires auth."
  def get_download_url(token, id, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getDownloadUrl", %{"id" => id})

  @doc "app.rocksky.library.getCoverArtUrl — requires auth."
  def get_cover_art_url(token, id, opts \\ %{}, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getCoverArtUrl", Map.merge(%{"id" => id}, opts))

  @doc "app.rocksky.library.getInternetRadioStations — requires auth."
  def get_internet_radio_stations(token, base \\ ""),
    do: :rocksky.library_get(base, token, "app.rocksky.library.getInternetRadioStations", %{})
end

File: sdk/go/rocksky/library.go
package rocksky

import (
	"context"
	"encoding/json"
	"errors"

	"github.com/bluesky-social/indigo/xrpc"
)

// Library is the authenticated client for app.rocksky.library.* — the Subsonic /
// navidrome-compatible surface over a user's uploaded music. Every method
// requires auth, so a Library can only be obtained from [Client.Library], which
// errors unless a token was attached via [Client.WithToken].
//
// Outputs are the AppView's raw JSON payloads (the library lexicons are
// intentionally loose), returned as json.RawMessage. Methods that take optional
// parameters accept them as an opts map keyed by the camelCase field name (pass
// nil for none).
type Library struct {
	xrpc *xrpc.Client
}

// Library returns the authenticated app.rocksky.library.* client. It errors
// unless an access token was attached via [Client.WithToken].
func (c *Client) Library() (*Library, error) {
	if c.xrpc.Auth == nil || c.xrpc.Auth.AccessJwt == "" {
		return nil, errors.New("app.rocksky.library.* requires an access token; call WithToken first")
	}
	return &Library{xrpc: c.xrpc}, nil
}

func (l *Library) query(ctx context.Context, nsid string, params map[string]any) (json.RawMessage, error) {
	filtered := make(map[string]any, len(params))
	for k, v := range params {
		if s, ok := v.(string); ok && s == "" {
			continue
		}
		filtered[k] = v
	}
	var out json.RawMessage
	if err := l.xrpc.Do(ctx, xrpc.Query, "", nsid, filtered, nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

func (l *Library) procedure(ctx context.Context, nsid string, body map[string]any) (json.RawMessage, error) {
	var out json.RawMessage
	if err := l.xrpc.Do(ctx, xrpc.Procedure, "application/json", nsid, nil, body, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// Ping calls app.rocksky.library.ping (requires auth).
func (l *Library) Ping(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.ping", nil)
}

// GetLicense calls app.rocksky.library.getLicense (requires auth).
func (l *Library) GetLicense(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getLicense", nil)
}

// GetMusicFolders calls app.rocksky.library.getMusicFolders (requires auth).
func (l *Library) GetMusicFolders(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getMusicFolders", nil)
}

// GetScanStatus calls app.rocksky.library.getScanStatus (requires auth).
func (l *Library) GetScanStatus(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getScanStatus", nil)
}

// StartScan calls app.rocksky.library.startScan (requires auth).
func (l *Library) StartScan(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.startScan", nil)
}

// GetUser calls app.rocksky.library.getUser (requires auth).
func (l *Library) GetUser(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getUser", nil)
}

// GetArtists calls app.rocksky.library.getArtists (requires auth).
func (l *Library) GetArtists(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getArtists", nil)
}

// GetIndexes calls app.rocksky.library.getIndexes (requires auth).
func (l *Library) GetIndexes(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getIndexes", nil)
}

// GetArtist calls app.rocksky.library.getArtist (requires auth).
func (l *Library) GetArtist(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.query(ctx, "app.rocksky.library.getArtist", params)
}

// GetArtistInfo calls app.rocksky.library.getArtistInfo (requires auth).
func (l *Library) GetArtistInfo(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.query(ctx, "app.rocksky.library.getArtistInfo", params)
}

// GetAlbum calls app.rocksky.library.getAlbum (requires auth).
func (l *Library) GetAlbum(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.query(ctx, "app.rocksky.library.getAlbum", params)
}

// GetAlbumList calls app.rocksky.library.getAlbumList (requires auth).
func (l *Library) GetAlbumList(ctx context.Context, typ string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["type"] = typ
	for k, v := range opts {
		params[k] = v
	}
	return l.query(ctx, "app.rocksky.library.getAlbumList", params)
}

// GetAlbumInfo calls app.rocksky.library.getAlbumInfo (requires auth).
func (l *Library) GetAlbumInfo(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.query(ctx, "app.rocksky.library.getAlbumInfo", params)
}

// GetSong calls app.rocksky.library.getSong (requires auth).
func (l *Library) GetSong(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.query(ctx, "app.rocksky.library.getSong", params)
}

// GetRandomSongs calls app.rocksky.library.getRandomSongs (requires auth).
func (l *Library) GetRandomSongs(ctx context.Context, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	for k, v := range opts {
		params[k] = v
	}
	return l.query(ctx, "app.rocksky.library.getRandomSongs", params)
}

// GetSongsByGenre calls app.rocksky.library.getSongsByGenre (requires auth).
func (l *Library) GetSongsByGenre(ctx context.Context, genre string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["genre"] = genre
	for k, v := range opts {
		params[k] = v
	}
	return l.query(ctx, "app.rocksky.library.getSongsByGenre", params)
}

// GetSimilarSongs calls app.rocksky.library.getSimilarSongs (requires auth).
func (l *Library) GetSimilarSongs(ctx context.Context, id string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	for k, v := range opts {
		params[k] = v
	}
	return l.query(ctx, "app.rocksky.library.getSimilarSongs", params)
}

// GetTopSongs calls app.rocksky.library.getTopSongs (requires auth).
func (l *Library) GetTopSongs(ctx context.Context, artist string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["artist"] = artist
	for k, v := range opts {
		params[k] = v
	}
	return l.query(ctx, "app.rocksky.library.getTopSongs", params)
}

// GetLyrics calls app.rocksky.library.getLyrics (requires auth).
func (l *Library) GetLyrics(ctx context.Context, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	for k, v := range opts {
		params[k] = v
	}
	return l.query(ctx, "app.rocksky.library.getLyrics", params)
}

// GetMusicDirectory calls app.rocksky.library.getMusicDirectory (requires auth).
func (l *Library) GetMusicDirectory(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.query(ctx, "app.rocksky.library.getMusicDirectory", params)
}

// GetGenres calls app.rocksky.library.getGenres (requires auth).
func (l *Library) GetGenres(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getGenres", nil)
}

// Search calls app.rocksky.library.search (requires auth).
func (l *Library) Search(ctx context.Context, query string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["query"] = query
	for k, v := range opts {
		params[k] = v
	}
	return l.query(ctx, "app.rocksky.library.search", params)
}

// GetStarred calls app.rocksky.library.getStarred (requires auth).
func (l *Library) GetStarred(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getStarred", nil)
}

// Star calls app.rocksky.library.star (requires auth).
func (l *Library) Star(ctx context.Context, id string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	for k, v := range opts {
		params[k] = v
	}
	return l.procedure(ctx, "app.rocksky.library.star", params)
}

// Unstar calls app.rocksky.library.unstar (requires auth).
func (l *Library) Unstar(ctx context.Context, id string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	for k, v := range opts {
		params[k] = v
	}
	return l.procedure(ctx, "app.rocksky.library.unstar", params)
}

// GetPlaylists calls app.rocksky.library.getPlaylists (requires auth).
func (l *Library) GetPlaylists(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getPlaylists", nil)
}

// GetPlaylist calls app.rocksky.library.getPlaylist (requires auth).
func (l *Library) GetPlaylist(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.query(ctx, "app.rocksky.library.getPlaylist", params)
}

// CreatePlaylist calls app.rocksky.library.createPlaylist (requires auth).
func (l *Library) CreatePlaylist(ctx context.Context, name string) (json.RawMessage, error) {
	params := map[string]any{}
	params["name"] = name
	return l.procedure(ctx, "app.rocksky.library.createPlaylist", params)
}

// UpdatePlaylist calls app.rocksky.library.updatePlaylist (requires auth).
func (l *Library) UpdatePlaylist(ctx context.Context, playlistId string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["playlistId"] = playlistId
	for k, v := range opts {
		params[k] = v
	}
	return l.procedure(ctx, "app.rocksky.library.updatePlaylist", params)
}

// DeletePlaylist calls app.rocksky.library.deletePlaylist (requires auth).
func (l *Library) DeletePlaylist(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.procedure(ctx, "app.rocksky.library.deletePlaylist", params)
}

// DeleteSong calls app.rocksky.library.deleteSong (requires auth).
func (l *Library) DeleteSong(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.procedure(ctx, "app.rocksky.library.deleteSong", params)
}

// DeleteAlbum calls app.rocksky.library.deleteAlbum (requires auth).
func (l *Library) DeleteAlbum(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.procedure(ctx, "app.rocksky.library.deleteAlbum", params)
}

// Scrobble calls app.rocksky.library.scrobble (requires auth).
func (l *Library) Scrobble(ctx context.Context, id string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	for k, v := range opts {
		params[k] = v
	}
	return l.procedure(ctx, "app.rocksky.library.scrobble", params)
}

// UpdateNowPlaying calls app.rocksky.library.updateNowPlaying (requires auth).
func (l *Library) UpdateNowPlaying(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.procedure(ctx, "app.rocksky.library.updateNowPlaying", params)
}

// GetNowPlaying calls app.rocksky.library.getNowPlaying (requires auth).
func (l *Library) GetNowPlaying(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getNowPlaying", nil)
}

// GetPlayQueue calls app.rocksky.library.getPlayQueue (requires auth).
func (l *Library) GetPlayQueue(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getPlayQueue", nil)
}

// SavePlayQueue calls app.rocksky.library.savePlayQueue (requires auth).
func (l *Library) SavePlayQueue(ctx context.Context, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	for k, v := range opts {
		params[k] = v
	}
	return l.procedure(ctx, "app.rocksky.library.savePlayQueue", params)
}

// GetStreamUrl calls app.rocksky.library.getStreamUrl (requires auth).
func (l *Library) GetStreamUrl(ctx context.Context, id string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	for k, v := range opts {
		params[k] = v
	}
	return l.query(ctx, "app.rocksky.library.getStreamUrl", params)
}

// GetDownloadUrl calls app.rocksky.library.getDownloadUrl (requires auth).
func (l *Library) GetDownloadUrl(ctx context.Context, id string) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	return l.query(ctx, "app.rocksky.library.getDownloadUrl", params)
}

// GetCoverArtUrl calls app.rocksky.library.getCoverArtUrl (requires auth).
func (l *Library) GetCoverArtUrl(ctx context.Context, id string, opts map[string]any) (json.RawMessage, error) {
	params := map[string]any{}
	params["id"] = id
	for k, v := range opts {
		params[k] = v
	}
	return l.query(ctx, "app.rocksky.library.getCoverArtUrl", params)
}

// GetInternetRadioStations calls app.rocksky.library.getInternetRadioStations (requires auth).
func (l *Library) GetInternetRadioStations(ctx context.Context) (json.RawMessage, error) {
	return l.query(ctx, "app.rocksky.library.getInternetRadioStations", nil)
}
